### PR TITLE
Intervals builder

### PR DESCRIPTION
**Pull request by kbabushkin**
The PR introduces the value interval builder from the AST expression for a given parameter and specified comparison operations. More tests will be added later.

### CODE REVIEW

File: pkg/intervals/intervals.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
+package intervals
```
Comment (dspasibenko) [inline pkg/intervals/intervals.go:1]: Here and in other files let's add the license comment...

File: pkg/intervals/intervals.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+		Max  T                  // max value of the type T
+		CmpF func(v1, v2 T) int // comparison function of T type values
+	}
+
+	// Interval represents an interval of values of type T,
+	// interval can be open, half-open, closed.
+	Interval[T any] struct {
+		L   T
+		R   T
+		B   Basis[T]
```
Comment (dspasibenko) [inline pkg/intervals/intervals.go:25]: What is the reason that every interval object has the basis? Why not to build a set of intervals for the same basis, so the basis is the parameter of the Intervals builder?

**Review by dspasibenko — COMMENTED** [review]
I wish I understand the reason for "open" and "close" concept. Is it an attempt to avoid next(t) and prev(t) functions?

File: pkg/intervals/intervals.go (hunk 1 above)
Comment (kbabushkin) [reply pkg/intervals/intervals.go:15]: applied

File: pkg/intervals/intervals.go (hunk 2 above)
Comment (kbabushkin) [reply pkg/intervals/intervals.go]: Basis now produces the intervals and contains all the operations to handle intersection, union, etc.

**Review by dspasibenko — APPROVED** [review]

**Pull request merged**